### ISSUE TITLE
Search path for libraries (path spec v3)

### ISSUE THREAD

**Issue by cameel** [ethereum/solidity#11409]
Depends on #11408.
Related to #9353.
Part of a set of issues that replaces #11138.

## Abstract
Add `--include-paths` option which specifies a list of directories from which files can be loaded using relative paths (in addition to `--base-path`). This is meant to be similar to `$PATH` variable in the shell or `$PYTHONPATH` in Python.

The new feature makes it possible to reliably use relative paths in imports without relying on remappings. This is the first step towards disallowing absolute import paths, though for now it will still be possible to use an empty path as `--base-path` (in fact it remains the default) to keep using them.

## Motivation
Currently the compiler provides the `--base-path` option, which is a way to specify how import paths should be resolved by the default file loader. By default base path is empty, which means that relative paths are relative to the current working directory while absolute paths remain absolute. This makes imports change their meaning depending on the directory from which the compiler is invoked. Specifying a custom base path, on the other hand, results in both relative and absolute paths being interpreted as relative to the base directory. This prevents the file loader from accessing files outside of that directory and thus breaks the common usage pattern where libraries are installed using a package manager and can reside in different locations depending on how they were installed (e.g. local vs global `node_nodules/`).

From what I have seen, frameworks generally ignore `--base-path` and have their own mechanisms for dealing with importing libraries: they either don't rely on the default file loader at all (Truffle, Hardhat) or use import remappings to redirect relative paths to the right directories (Brownie, dapp.tools). The goal of this proposal is to extend the base path mechanism to be actually usable in direct CLI usage and optionally also for those frameworks that do rely on the default file loader. It might may also make it easier to use LSP with projects that rely on frameworks that do path resolution on their own (https://github.com/ethereum/solidity/issues/7763#issuecomment-832008294).

## Specification
1. Add a command-line option called `--include-path`.
    - The value is a single directory path. The option can be used multiple times to specify multiple directories.
    - The path must not be empty and must point at an existing directory.
    - The path is made canonical according to the same rules as `--base-path` (see #11408).
    - The option is allowed in combination with `--standard-json`, just like `--base-path` currently is.
2. When the default file loader receives a source unit name to load, it first tries to prepend base path to it and see if a file exists under that path. Then it does the same with all directories passed to `--include-path` in the order they were specified.
    - The loader returns the content of the first matching file.
    - If there is more than one matching file, the compiler prints a warning.
    - **NOTE**: This is different from how some frameworks handle this. For example Truffle's path resolver stops at the first match so it's possible for a file in one library to "shadow" a file in another.
3. Paths from `--include-paths` are automatically added to `--allowed-paths`.
4. Include paths are stripped from CLI file paths the same way `--base-path` is (this part depends on #11408).
5. Implement the same logic in [solcjs](https://github.com/ethereum/solc-js/blob/master/solcjs).

As a result base path simply becomes the first entry on a list of directories that are searched for import files. The only difference between it and other include paths is that it can be empty (and that difference will be removed by #11410).

**NOTE**: The paths specified in `sources.urls` in Standard JSON are also processed by the default file loader so they will also be affected by the new option.

## Backwards Compatibility
The change is fully backwards-compatible. Without using `--include-paths` we get the current behavior.

**Comment by chriseth** [comment]
Can you clarify the relation between base path and include path? Is `--base-path X --include-paths Y:Z` equivalent to `--include-paths X:Y:Z`?

Also, do we really want to use `:` as a separator? This could be hard on windows. What about using multiple `--include-paths` options instead?

**Comment by cameel** [comment]
> Can you clarify the relation between base path and include path? Is `--base-path X --include-paths Y:Z` equivalent to `--include-paths X:Y:Z`?

Base path basically becomes the first include path. The only difference between it and other include paths is that it can be empty.

It makes it almost redundant and we could remove it if we really wanted to but keeping it makes things backwards-compatible. Anyone already using `--base-path` today will be able to keep using it without any changes. I think it's also more convenient that way - you do not have to remember to include current working directory as the first entry in `--include-paths` (and this is what user will want in most cases).

> Also, do we really want to use `:` as a separator? This could be hard on windows. What about using multiple `--include-paths` options instead?

Good point. Forgot about paths that include drive letter. Yeah. I think requiring use to specify the option multiple times is a good alternative.

**Comment by cameel** [comment]
Updated the description. `--include-paths` now accepts a single directory and can be used multiple times. It's canonicalized the same way as `--base-path`.

I also removed the remark about disallowing duplicates. You could have something like `solc -I project/ -I /home/user/project/` and it would suddenly fail when you go to `/home/user/` because now both `project/`s are the same directory.

**Comment by cameel** [comment]
A change after today's call: the compiler should not stop after finding a file in one of the include directories and instead report a warning if there's another match in a different include directory.

**Comment by axic** [comment]
Warning or error? I think we were leaning towards error for the beginning.

**Comment by cameel** [comment]
I think @chriseth in the end said that we should print something but still proceed using the first file found.

**Comment by cameel** [comment]
A, I see what you mean. I wrote "warning" in the comment but still said that it should fail in the description. I'll fix that :)

**Comment by cameel** [comment]
#11848 implemented the `solc` part of the feature. There are some minor changes compared to the spec above:
- There's no short version of the option. @axic has pointed out in another PR some time ago that we do not have them on purpose.
- If there is more than one file matching an import, the callback reports an error rather than a warning. The callback has no mechanism for reporting warnings.
- The include dirs do not have to exist.
- Extra limitation: base path must be non-empty if `--include-path` is used.

**Comment by cameel** [comment]
solc-js part implemented in https://github.com/ethereum/solc-js/pull/544.

Both solc-js and solc PRs have been merged so I'm closing this.
